### DISCUSSION TITLE
How to find helpful features?

### DISCUSSION THREAD

**Discussion by RuelYasa** [bevyengine/bevy#14095] [Q&A]
﻿I readed Bevy's changelogs recently, and found many new features I havn't known before, which are beneficial if used in projects. However, they are too many to be entirely memorized for me, and are unknown to who hasn't readed the changelogs. Also, their number will continuously increase in the future, so it's impractical to remember them all.

Therefore, having a method to browse features is necessary. I usually read examples to explore features, but not every feature is included in the examples. Another possible way is to read documents, but it is very inefficient, and cannot highlight appealing features, because there is no application scenario, advantage or example in the documents, so it is hard to distinguish needed features from other functions.
﻿
So, I wonder how to spot helpful features from abundant functions. For example:Imagine that I am programming a game which loads many assets when opened, which has caused performance issues, but I know nothing about bevy's asset preprocessing. Then, how can I get the inspiration to use it to improve performance? Is there a "list of features" that I can read through to discover the preprocessing? Or the only way is memorizing all the features, and keeping up with every updates?

**Comment by JMS55** [comment]
Ideally this would be documented in the module docs of each crate, but in practice we don't have great documentation.

**Comment by bugsweeper** [comment]
This was an issue before 0.14 version, starting from 0.14 all gated by feature types have concrete mention of feature like [here](https://dev-docs.bevyengine.org/bevy/render/texture/fn.get_transcoded_formats.html)
![image](https://github.com/bevyengine/bevy/assets/17225606/30196397-2075-4a34-8741-4f893ddd5f2a)
And a lot of work still should be done, but we are [moving forward](https://github.com/bevyengine/bevy/pull/13712)

**Reply by JMS55** [reply]
It's a bit ambiguous, but I think the OP meant features as a general term, and not specifically cargo `features`.

**Reply by bugsweeper** [reply]
Ah-ah-ah, in this sense I would look at articles of fresh releases and examples. I think examples would be even better, but we should improve it (add more feature aligned examples, add support for animated preview in example list page)
![image](https://github.com/bevyengine/bevy/assets/17225606/31126c31-7c3b-4523-aa56-57ac1c683193)

**Reply by RuelYasa** [reply]
> Ah-ah-ah, in this sense I would look at articles of fresh releases and examples. I think examples would be even better, but we should improve it (add more feature aligned examples, add support for animated preview in example list page) 


It is possible, but as what I have mentioned, referring to articles whose number is keep growing is seemingly not a sustainable solution, because it will be even harder to remember them all in the further, and it is hard to locate the useful category, rather than be flooded in unneeded rendering improvements. Therefore, a comprehensive and organized "reference book" is necessary.
﻿
Examples are indeed good documents, but what we need to know is mostly the existence of the features rather than how to use them, which makes examples a bit cumbersome.
﻿
In my opinion, if extra efforts can be made, a reference in the form like mdBook with each feature as a title would be a elegant solution, because users can simply read through the index to get informed of all features, without needing to do extra operations. The only problem is that I shamefully have not engaged in any big project like bevy, so I do not know how much effort is needed to maintain such a reference.
